File: design-pattern-java/app/src/main/java/design/pattern/java/patterns/structural/composite/ContainerComposite.java
package design.pattern.java.patterns.structural.composite;

import java.util.ArrayList;
import java.util.List;

public class ContainerComposite implements UIComponent {
    private List<UIComponent> children;
    public ContainerComposite(UIComponent component) {
        this.children = new ArrayList<>();
        this.children.add(component);
    }
    @Override
    public String render() {
        return children.stream()
                .map(component -> component.render())
                .reduce("",
                        (acc, current) -> acc.concat(" ")
                                .concat(current));
    }
    public ContainerComposite add(UIComponent component) {
        this.children.add(component);
        return this;
    }
    public void remove(UIComponent component) {
        this.children.remove(component);
    }
}

File: design-pattern-jvm/app/src/main/kotlin/design/pattern/kotlin/patterns/structural/composite/ContainerComposite.kt
package design.pattern.kotlin.patterns.structural.composite
class ContainerComposite(component: UIComponent) : UIComponent {
    private val children: MutableList<UIComponent> = mutableListOf()
    init {
        children.add(component)
    }
    override fun render(): String {
        return children.stream()
            .map { component: UIComponent -> component.render() }
            .reduce(
                ""
            ) { acc: String, current: String -> "$acc $current" }
    }

    fun add(component: UIComponent): ContainerComposite {
        children.add(component)
        return this
    }

    fun remove(component: UIComponent) {
        children.remove(component)
    }


}